src/App.js

```diff
@@ -28,25 +31,91 @@ const generateSquares = () => {
+        }
+      }
+      return squareArray;
+    });
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+    if (player === PLAYER_1) { // logic to toggle which player's turn it is, which changes whether to set the square's value as 'x' or 'o'
+      playerTurn(PLAYER_2);
+    }
+    else {
+      playerTurn(PLAYER_1);
+    }
+    setSquares(newSquares); // the new state of squares is in newSquares, which is now to be updated by setSquares
+    checkWinner(checkForWinner()); // checks/updates state of winner, and calls checkForWinner() inside checkWinner
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    // this function first checks row by row whether the value in each square is the same; if it is, we return that value to checkWinner() and update the state of winner
+    // then checks column by column
+    // then checks for a winner diagonally
+
+    let row = 0;
+    let col = 0;
+
+    // check for winning combo in rows
+    while (row < 3) {
+      if (
+        squares[row][0].value === squares[row][1].value &&
+        squares[row][2].value === squares[row][1].value &&
+        squares[row][0].value != ''
+      ) {
+        return squares[row][0].value;
+      } else row += 1;
+    }
+    // check for winning combo in columns
+    while (col < 3) {
+      if (
+        squares[0][col].value === squares[1][col].value &&
+        squares[2][col].value === squares[1][col].value &&
+        squares[0][col].value != ''
+      ) {
+        return squares[0][col].value;
+      } else col += 1;
+    }
+    // check for winning combo diagonally
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[0][0] != ''
+    ) {
+      return squares[0][0].value;
+    } else if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[0][2] != ''
+    ) {
+      return squares[0][2].value;
+    } else { // if we don't find a winner in the rows, columns, or diagonally, we look for squares the value of '', which means the game is ongoing, and the winner still has state of ''
+      for (let array of squares) {
+        for (let element of array) {
+          if (element.value == '') {
+            return '';
+          }
+          continue;
+        }
+      }
+      return 'Tie'; // if all squares have a value (aka value is not ''), the game is tied, and the state of winner is 'Tie'
+    }
   };
 
+
   const resetGame = () => {
```

Comment:
To reset the game, all of the state variables must be reset to their starting values.  This can be done by calling all of the set state functions and passing in the initial default values as the new values for each state variable.